web3/providers/persistent/subscription_manager.py

```diff
@@ -0,0 +1,151 @@
+                    "Subscription label already exists. Subscriptions must have "
+                    f"unique labels.\n    label: {subscriptions.label}"
+                )
+
+            subscriptions.manager = self
+            sx_id = await self._w3.eth._subscribe(*subscriptions.subscription_params)
+            subscriptions._id = sx_id
+            self._subscriptions_by_label[subscriptions.label] = subscriptions
+            self._subscriptions_by_id[subscriptions.id] = subscriptions
+            self.subscriptions.append(subscriptions)
+            self.logger.info(
+                "Successfully subscribed to subscription:\n    "
+                f"label: {subscriptions.label}\n    id: {sx_id}"
+            )
+            return sx_id
+        elif isinstance(subscriptions, Sequence):
+            if len(subscriptions) == 0:
+                raise Web3ValueError("No subscriptions provided.")
+
+            sx_ids: List[HexStr] = []
+            for sx in subscriptions:
+                await self.subscribe(sx)
+            return sx_ids
+        raise Web3TypeError("Expected a Subscription or a sequence of Subscriptions.")
+
+    async def unsubscribe(self, subscription: EthSubscription[Any]) -> bool:
+        """
+        Used to unsubscribe from a subscription that is being managed by the
+        subscription manager.
+
+        :param subscription: The subscription to unsubscribe from.
+        :type subscription: EthSubscription
+        :return: `True` if unsubscribing was successful, `False` otherwise.
+        :rtype: bool
+        """
+        if subscription not in self.subscriptions:
+            raise Web3ValueError(
+                "Subscription not found or is not being managed by the subscription "
+                f"manager.\n    label: {subscription.label}\n    id: {subscription._id}"
+            )
+        if await self._w3.eth._unsubscribe(subscription.id):
+            self.subscriptions.remove(subscription)
+            self._subscriptions_by_id.pop(subscription.id)
+            self._subscriptions_by_label.pop(subscription.label)
+            self.logger.info(
+                "Successfully unsubscribed from subscription:\n    "
+                f"label: {subscription.label}\n    id: {subscription.id}"
+            )
+            return True
+        return False
+
+    async def unsubscribe_all(self) -> None:
+        """
+        Used to unsubscribe from all subscriptions that are being managed by the
+        subscription manager.
+
+        :return: None
+        """
+        unsubscribed = [await self.unsubscribe(sx) for sx in self.subscriptions.copy()]
+        if all(unsubscribed):
+            self.logger.info("Successfully unsubscribed from all subscriptions.")
+        else:
+            if len(self.subscriptions) > 0:
+                self.logger.warning(
+                    "Failed to unsubscribe from all subscriptions. Some subscriptions "
+                    "are still active."
```

Comment:
Do you think there is a benefit in keeping both the debug log and the warning log here?

Reply:
Debug includes the warning as well, but if you only set the log level to warning (warning and errors only) you will still see this log. Happy to remove though, thoughts?

Reply:
Maybe putting the debug content (the subscriptions that are still present) into the warning? 🤔